sway-core/src/lib.rs

```diff
@@ -438,22 +439,21 @@ fn is_parse_module_cache_up_to_date(
             // we got from the cache, which is only true if the file hasn't been
             // modified since or if its hash is the same.
             let cache_up_to_date = entry.modified_time == modified_time || {
-                let src = std::fs::read_to_string(path.as_path()).unwrap();
-
                 let mut hasher = DefaultHasher::new();
-                src.hash(&mut hasher);
+                // It's important to hash the input that triggered compilation. We don't want to
+                // load the file from disk as it may have been changed by LSP or another process since
+                // compilation started.
+                input.hash(&mut hasher);
                 let hash = hasher.finish();
-
                 hash == entry.hash
             };
 
             // Look at the dependencies recursively to make sure they have not been
             // modified either.
             if cache_up_to_date {
-                entry
-                    .dependencies
-                    .iter()
-                    .all(|path| is_parse_module_cache_up_to_date(engines, path, include_tests))
+                entry.dependencies.iter().all(|path| {
+                    is_parse_module_cache_up_to_date(engines, path, input.clone(), include_tests)
```

Comment:
Good spot, I've added a new GC test for projects with multiple files and it's not passing now. I'll have a think about both of the points above to see what makes the most sense. Will put the PR back into draft until this new test passes.

Reply:
Just an update, I have a fix for this but it's not very pretty in it's current shape. Will tidy and push tomorrow. Managed to fix another bug at the same time so we now no longer need to clone the query engine also.

Reply:
Ok, I'm now passing through the document version that is supplied by the LSP client. Otherwise we fallback to the previous technique which is safe when not being edited by LSP.